Changes for running Zephyr as Xen control domain

firscity opened the pull request:
Currently, Zephyr RTOS can be run as Domain-U (unprivileged domain) under Xen hypervisor via xenvm board. But Zephyr looks like great safe Linux alternative for Xen Domain-0 (privileged domain).

Zephyr kernel now includes lot of required drivers (Xen events, grant tables etc.), but start as a control domain (Domain-0) requires additional changes: domain control interface, possibility to configure any Zephyr board to run as Xen control domain etc. They were added in this PR.

Patch `arch: arm64: mmu: WA: avoid using of set/way cache instructions for Xen` is only for demonstration/testing purposes and will be replaced with proper fix. Related discussion about set/way problem under Xen hypervisor: #58436.

firscity committed:
xen: dom0: update Xen public headers for running Zephyr as Dom0

Xen public headers are added to Zephyr kernel partially, so we need
to update them for new features implementation.

Further implementation of Xen domains configuration and memory
management requires dedicated Xen public headers inside Zephyr kernel.
It will be used for add Zephyr OS Xen Domain-0 funtionalities.

Update existent and add new required public headers from Xen 4.17.0
release.

Signed-off-by: Dmytro Firsov <dmytro_firsov@epam.com>

firscity committed:
xen: change HVM functions signature to run it for other domains

This commit adds possibility to call hypervisor HVM parameter functions
for specified domain (instead of only DOMID_SELF). It is needed for
configuring domains, that were created from Zephyr control domain.

Signed-off-by: Dmytro Firsov <dmytro_firsov@epam.com>

firscity committed:
xen: events: add event channel allocation for domain-0

This commit adds interface for evtchn allocation that can be used by
privilaged domain. Domain 0 can specify both dom and remote_dom
parameters for hypercall, where in others domains dom should be always
DOMID_SELF. It is needed for creating pre-defined channels during
domain setup in Zephyr Dom0.

Signed-off-by: Dmytro Firsov <dmytro_firsov@epam.com>

xakep-amatop committed:
dts: bindings: xen: add xen,xen.yaml file

Add yaml file for 'xen,xen', because without it an appropriate
'CONFIG_DT_HAS_XEN_XEN_ENABLED' isn't generated.

It will be used for checking Xen support on current setup, instead of
checking if it is BOARD/SOC "xenvm" (which is not correct for Domain-0
configurations).

Remove xen,xen-4.15.yaml at all, because it isn't necessary to have
yaml for some specific Xen version.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>
Signed-off-by: Dmytro Firsov <dmytro_firsov@epam.com>

xakep-amatop committed:
drivers: xen: gnttab: do Xen node mapping inside driver

Move memory mapping of Xen node to Grant Table driver system init
function. After moving mapping we don't need anymore records of
xen-xen node into 'mmu_regions' array, so they were deleted from
all SoCs: Rcar Gen3/Gen4 and XenVM.

We need at least 16M of virtual address space to map memory of Xen
node, so the virtual memory sized has been increased to 32 MB, it
should be enough for basic use-cases and mapping of 16M mem region
of Xen node.

Unfortunately, after moving we also need to increase number of XLAT
tables. The previous code was more efficient if we talking about
usage of XLAT tables, because it mapped grant tables using a higher-
order table that allows mapping blocks of 2MB. And after the changes
is maps every 4KB page, so we need more XLAT tables.

Increase number of grant frames, it is needed to sync stage 1 and stage 2
memory mappings, previously we map only one page on stage 2 and further
usage of unmap regions can cause MMU translation errors.

Perform mapping stage 1 before mapping for stage 2 (add to physmap),
because right after stage 1 we can try to access memory and if it is
unmap in stage 2, error will be received during translation.

Note: Xen Grant Table driver doesn't use Zephyr Device Model.

Authored-by: Mykola Kvach <mykola_kvach@epam.com>
Co-authored-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>
Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>

firscity committed:
xen: implement Xen domain control interface

Add Xen domctl API implementation for Zephyr as control domain.

Previously Zephyr OS was used as unprivileged Xen domain (Domain-U),
but it also can be used as lightweight Xen control domain (Domain-0).
To implement such fuctionality additional Xen interfaces are needed.
One of them is Xen domain controls (domctls) - it allows to create,
configure and manage Xen domains.

Also, used it as a possibility to update files copyright and licenses
identifiers in touched files.

Signed-off-by: Dmytro Firsov <dmytro_firsov@epam.com>

firscity committed:
xen: add helper functions for Xen domain memory management

Add wrapper functions for Xen memory managment hypercall. They can be
used for unprivilaged Zephyr guest initialization and for domain
management, when Zephyr is used as Xen Domain 0.

Signed-off-by: Dmytro Firsov <dmytro_firsov@epam.com>

firscity committed:
snippets: add virtual Xen dom0 snippet support

Zephyr snippets allow to apply pre-defined build configuration and
extentions to your build0. It is made by applying generic snippet
conf/device-tree changes and may be extended with platform-specific
files.

It looks like great approach for convertion Zephyr OS build to Xen
control domain on different boards, as it requires Kconfig changes
and board device tree changes (configuration for memory node, hypervisor
console, grant table region etc).

Please note, that Xen hypervisor passes all selected parameters via
domain device-tree, but Zephyr does not use it in runtime. So, user
should keep values in board overlays in sync with real Xen values.

Add example and docs for converting qemu_cortex_a53 board to Xen
Domain-0 guest. This approach might be used for other boards, where
Zephyr Dom0 is needed.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>
Signed-off-by: Dmytro Firsov <dmytro_firsov@epam.com>

carlocaione commented on drivers/CMakeLists.txt:
```diff
@@ -12,7 +12,7 @@ add_subdirectory(usb_c)
 add_subdirectory_ifdef(CONFIG_AUDIO audio)
 add_subdirectory_ifdef(CONFIG_BBRAM bbram)
-add_subdirectory_ifdef(CONFIG_BOARD_XENVM xen)
+add_subdirectory_ifdef(CONFIG_DT_HAS_XEN_XEN_ENABLED xen)
```
uhm, for consistency this should be `add_subdirectory_ifdef(CONFIG_XEN)` (or similar) and then use `CONFIG_DT_HAS_XEN_XEN_ENABLED` in the actual driver (see how this is done for the other drivers).

carlocaione reviewed (changes requested):
This is a bit out of my comfort zone, so just some trivial nits for now.

povergoing reviewed (commented):
Zephyr as DOM0 is really impressive, only some nits I comment inline

povergoing commented:
A quick question, does this DOM0 support SMP mode?

firscity committed:
xen: refactor Xen hypervisor Kconfig options

Xen-related Kconfig options were highly dependand on BOARD/SOC xenvm.
It is not correct because Xen support may be used on any board and SoC.

So, Kconfig structure was refactored, now CONFIG_XEN is located in
arch/ directory (same as in Linux kernel) and can be selected for
any Cortex-A arm64 setup (no other platforms are currently supported).

Also remove confusion in Domain 0 naming: Domain-0, initial domain,
Dom0, privileged domain etc. Now all options related to Xen Domain 0
will be controlled by CONFIG_XEN_DOM0.

Signed-off-by: Dmytro Firsov <dmytro_firsov@epam.com>

firscity commented on include/zephyr/xen/public/arch-arm.h:
```diff
@@ -205,15 +205,138 @@ typedef uint64_t xen_pfn_t;
+
+	struct vcpu_guest_core_regs user_regs; /* Core CPU registers */
+
+	uint64_t sctlr;
```
sdf

firscity commented:
First of all, @carlocaione @povergoing thank you for review and fresh look on these changes.

To prevent mail spam (can't reply once with "Start review" on "outdated" comments), here are a quoted replies:


> What about instead moving this somewhere in arch/arm64/core? Having this in drivers is odd.

>no need to have default n.

> uhm, for consistency this should be add_subdirectory_ifdef(CONFIG_XEN) (or similar) and then use CONFIG_DT_HAS_XEN_XEN_ENABLED in the actual driver (see how this is done for the other drivers).

> Are the depends on SHIELD_XEN_DOM0 and depends on XEN_INITIAL_DOMAIN somehow duplicated?
Any case is there that XEN_INITIAL_DOMAIN is true but SHIELD_XEN_DOM0 is false?

> Do INITIAL_DOMAIN and DOM0 mean the same concept? If they are the same, sometimes INITIAL_DOMAIN and sometimes DOM0 are a bit weird.
...

These comments are completely right, we needed a fresh look on what is already done. I refactored this part and removed all redundant configs with just `CONFIG_XEN` and `CONFIG_XEN_DOM0`. It's enough to determine a type of Xen setup and can be easily selected when "xen,xen" node is present (`XEN_DOM0` - when node is present and shield is applied). Thank you, fixed.


> this looks weirdly indented (like the rest of the file).

> Keep spacing consistent.

> BIT() ?

This file is just a copy of original Xen public header. I fixed coding style, but left bit values definitions as is (at some time it will probably require an update from fresh Xen version).


> cortex-a57?

Oh, sorry.  Sure, `cortex-a53`, fixed.


> A quick question, does this DOM0 support SMP mode?

We hadn't need to enable more than one core for both xenvm and Xen Dom0 setups, so I don't have an answer for that, needs testing.

carlocaione commented on arch/arm64/core/Kconfig:
```diff
@@ -158,6 +158,21 @@ config ARMV8_A
 	  so that it can support some features included in the AArch64 state.
 	  It supports the T32 and A32 instruction sets.
 
+config XEN
+	bool
+	default y
+	depends on DT_HAS_XEN_XEN_ENABLED
+	help
+	  Enables support of Xen hypervisor on arm64 platform. Get enabled
+	  when board device tree contains "hypervisor" node with "xen,xen"
+	  compatible enabled.
```
Oh man, this is super weird to have here.

Ok, it's time to have XEN as first class citizen then. Can you move all the XEN related stuff into a `xen/Kconfig.xen` file that you `rsource` from here? At least we do not pollute the general `Kconfig` too much.

firscity replied:
@carlocaione had such ideas too during refactor. I will fix it soon along with CI. Thank you.

firscity committed:
tests: kernel: exclude xenvm from device tests

Zephyr device and device.pm tests uses device tree fragments applied
to main board device trees. For xenvm they have conflicting
address/size cells definition with board DT. It leads to CI and test
issues during build (xenvm has 0x2 cells, tests have 0x1).

Signed-off-by: Dmytro Firsov <dmytro_firsov@epam.com>

luca-fancellu commented on drivers/xen/events.c:
```diff
@@ -47,6 +47,22 @@ int alloc_unbound_event_channel(domid_t remote_dom)
 	return rc;
 }
 
+int alloc_unbound_event_channel_dom0(domid_t dom, domid_t remote_dom)
```
Please correct me if I am wrong, but can this be protected with the dom0 Kconfig? Would it make sense?

luca-fancellu commented on drivers/xen/memory.c:
```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright (c) 2023 EPAM Systems
+ */
+#include <zephyr/arch/arm64/hypercall.h>
```
NIT: can a blank line be added before the include?

luca-fancellu reviewed (commented):
Nice work!

jgrall commented:
Commit message for "drivers: xen: gnttab: do Xen node mapping inside driver"

> Unfortunately, after moving we also need to increase number of XLAT
tables. The previous code was more efficient if we talking about
usage of XLAT tables, because it mapped grant tables using a higher-
order table that allows mapping blocks of 2MB.

From the commit message it is not entirely clear why you couldn't continue to use 2MB afterwards. That said, the original code was buggy because AFAIK we don't guarantee that the grant-table area will be 2MB aligned. So you may have mapped memory with the wrong attributes.

jgrall commented on arch/arm64/core/mmu.c:
```diff
@@ -816,7 +816,12 @@ static void enable_mmu_el1(struct arm_mmu_ptables *ptables, unsigned int flags)
 	barrier_isync_fence_full();
 
 	/* Invalidate all data caches before enable them */
+#ifdef CONFIG_DT_HAS_XEN_XEN_ENABLED
```
So there are a few issues with using set/way instructions in Zephyr:
  * As you said above they are not easily virtualizable
  * On Armv8, sytem caches are can only be cleaned by VA. So this would not flush all the caches.

If you want more details, there is a pretty good talk from Marc Zyngier about it:
https://www.youtube.com/watch?v=A_zCxzpxzmE

Therefore, I think whatever solution you come up with should be generic.

firscity replied:
Thank you for addition, we currently have discussion #58436 for this issue. We tried U-boot approach, but it will have issues during dynamic memory mapping. For sure, final solution should be generic, this commit is here only for test purposes during review.

xakep-amatop replied:
> Therefore, I think whatever solution you come up with should be generic.

Here is the proper fix instead of the current workaround: https://github.com/zephyrproject-rtos/zephyr/pull/61035

firscity commented:
@jgrall thank you for your review!

> What's your plan for other Xen versions and going forward?

Now we want to add support for 4.17 release, based on its fixed domctl interface. Starting from 4.18 it is planned to update it only for stable releases and separate changes basing on major/minor Xen version defined during build.

> From the commit message it is not entirely clear why you couldn't continue to use 2MB afterwards. That said, the original code was buggy because AFAIK we don't guarantee that the grant-table area will be 2MB aligned. So you may have mapped some memory with the wrong atributes (for instance if the rest of the 2MB is MMIO).

Yes, sure, regions may be unaligned to 2MB. Both previous and current approaches are correct, but works differently. Previously, Zephyr used 1:1 mapping and in `xenvm` case it was possible to use L2 because address was 2MB aligned. Now we are using `device_map()` which picks virtual region from internal allocator (`virt_region_alloc()`) and usually returned regions are 4K page aligned. It leads to mapping grant-table area with L3 entries and consuming more XLAT tables.
Unfortunately, Zephyr `device_map()` does not offer interface for specifying virt_region alignment during mapping.

jgrall commented:
> Now we want to add support for 4.17 release, based on its fixed domctl interface. Starting from 4.18 it is planned to update it only for stable releases and separate changes basing on major/minor Xen version defined during build.

I am a bit concerned of the complexity going forward. In the description, you suggested the goal is to provide a toolstack on top of Zephyr. Can you clarify your end goal? Is it creating domain based on a statically defined configuration?

firscity commented:
@jgrall 

> You will most likely need to provide wrapper that will abstract the difference in version from the callers.

I think it can be done in Zephyr lib level and lib should have relatively stable interface for user application and be synced with Zephyr kernel.

> Before adding the helpers in Zephyr, I think it would be good to clarify your end goal. You are saying you will implement a toolstack. But how powerful will it be? E.g. will it create domain dynamically or will they be based on a statically defined configuration (like dom0less)?

Currently we are working on wrapper library, that can be included in user application and allows to manage domains, start console and xenstore server, work with device tree etc. It can load configuration, domain kernels and device trees dynamically (e.g. from filesystem) or create domain with pre-defined configuration.
It's called `zephyr-xenlib` and located [here](https://github.com/xen-troops/zephyr-xenlib).

jfischer-no commented on boards/shields/xen_dom0/doc/index.rst:
```diff
@@ -0,0 +1,52 @@
+Overview
+********
+
+This virtual shield allows user to build Zephyr as a Xen initial domain (Dom0). The feature
```
The community does not like "virtual" shields, instead snippets should be used, even if they do not offer the same functionality. https://docs.zephyrproject.org/latest/build/snippets/index.html

xakep-amatop replied:
Just for clarification, when we were developing this functionality, the `snippets` concept didn't exist.
Thank you for pointing that out; we'll investigate this approach.

firscity commented:
@xakep-amatop provided proper fix for set/way issues in #61035, so this PR is blocked until set/way patches will be merged.

I will drop WA set/way patch and will try to rework Dom0 build configuration to snippets (thanks @jfischer-no for suggestion), after first look I see that it is possible.

xakep-amatop commented on snippets/xen_dom0/snippet.yml:
```diff
@@ -0,0 +1,9 @@
+  EXTRA_CONF_FILE: xen_dom0.conf
+
+boards:
+  qemu_cortex_a53:
```
you can add `rcar_h3ulcb_ca57` and `salvator_xs_m3` too, support of them have been added to mainline recently

firscity replied:
@xakep-amatop, added in separate commit.

firscity committed:
snippets: xen_dom0: add support for Renesas R-Car Gen3 boards

Renesas rcar_h3ulcb_ca57 and rcar_salvator_xs_m3 boards support recently
was added to Zephyr mainline. Add Xen control domain guest support for
them via xen_dom0 snippet extention.

Note: please pay attetion to overlay nodes and comments in case of any
issues on your setup.

Signed-off-by: Dmytro Firsov <dmytro_firsov@epam.com>

carlocaione reviewed (dismissed):
I mean, in general LGTM but again I'm way out my conform zone here. So giving my +1 but I think we need a XEN maintainer in Zephyr.

firscity commented:
@carlocaione, @xakep-amatop thank you for approval. Actually me and other my team member are Xen maintainers in Zephyr. These changes passed internal review.

But I will be very grateful if @jgrall will take a look at this and #61035 (blocks this PR), he is one of Xen maintainers for ARM64.

povergoing reviewed (dismissed):
LGTM

firscity commented:
Dear all, since #61035 was merged today, this PR is unblocked and can be merged (requested changes with snippets were implemented).

firscity commented:
@jfischer-no can you please re-check snippets implementation due to requested changes?

jfischer-no reviewed (dismissed):
LGTM, but maybe @mbolivar-ampere has time to take a look at the new snippet.

firscity commented:
Changes were rebased on top of `main` branch due to conflicts in testcase.yml file. No functional changes were made.

firscity commented:
In last force-push: fixed issues in `kernel.device` CI tests (add exclude for newly added due to same reason of invalid build configuration).

As previous approvals were dismissed, can @carlocaione @povergoing @jfischer-no check it again (no functional changes were made, only rebase)? Or it is ready to merge?

carlocaione commented:
@firscity can you fix the `WARNING: Title underline too short.` warning?

firscity commented:
@carlocaione oh, I'm sorry, missed this.
Fixed

mbolivar-ampere reviewed (commented):
@jfischer-no thanks for the ping. The snippet seems fine to me. At some point if this hits scalability limits we should consider adding the board directory to SNIPPET_ROOT automatically, to allow board support to be factored out from the main snippet implementation. But for now this seems fine. Let's see how it scales in practice.

carlocaione commented:
@firscity I do not understand why the documentation step is failing. Could you please try to rebase on the latest `main`?

carlocaione commented:
> @carlocaione, I already did this, but it did not helped, rebased once again. As I said previously, documentation builder incorrectly parses C struct in Xen header, which leads to a warning. And warning stops documentation build.

Ooh, ok, then this will need to be merged with superpowers.

povergoing reviewed (approved):
refresh +1

firscity commented:
@carlocaione can you, please, ask someone with superpower to merge this? All other CI test finished successfully.

carlocaione commented:
> @carlocaione can you, please, ask someone with superpower to merge this? All other CI test finished successfully.

@carlescufi @fabiobaltieri ^
